File: design/0745-prefix-and-suffix-search.py
import unittest
from typing import List

# https://leetcode.com/problems/prefix-and-suffix-search/

# python3 -m unittest design/0745-prefix-and-suffix-search.py

class TrieNode:
    def __init__(self):
        self.index = -1
        self.children = {}

class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word: str, index: int) -> None:
        curr = self.root
        for c in word:
            if c not in curr.children:
                curr.children[c] = TrieNode()
            curr = curr.children[c]
            curr.index = index

    def find(self, word: str) -> int:
        curr = self.root
        for c in word:
            if c not in curr.children: return -1
            curr = curr.children[c]
        return curr.index

class WordFilter:

    def __init__(self, words: List[str]):
        self.trie = Trie()
        for index, word in enumerate(words):
            word_len = len(word)
            word = word + "#" + word
            for i in range(word_len):
                self.trie.insert(word[i:], index)

    def f(self, pref: str, suff: str) -> int:
        word = suff + "#" + pref
        return self.trie.find(word)

class TestWordFilter(unittest.TestCase):
    def test(self) -> None:
        for commands, values, expected in [
            (["WordFilter", "f", "f", "f"], [["apple", "app", "ape"], ["a", "e"], ["appl", "pple"], ["apple", "apple"]], [None, 2, 0, 0]),
            (["WordFilter","f"], [["abbba","abba"],["ab","ba"]], [None,1]),
        ]:
            for idx, command in enumerate(commands):
                output = None
                match command:
                    case "WordFilter":
                        filter = WordFilter(values[idx])
                    case "f":
                        output = filter.f(values[idx][0], values[idx][1])
                self.assertEqual(expected[idx], output, f"expected: {expected[idx]}, output: {output}")

# Your WordFilter object will be instantiated and called as such:
# obj = WordFilter(words)
# param_1 = obj.f(pref,suff)

File: design/0745-prefix-and-suffix-search.go
package design

// https://leetcode.com/problems/prefix-and-suffix-search/

// Time: O(N∗M)
// Space: O(N∗M)
type WordFilter struct {
	index    int
	children map[byte]*WordFilter
}

func NewWordFilter(words []string) WordFilter {
	root := WordFilter{children: make(map[byte]*WordFilter)}

	var wordLen, newWordLen int

	var c byte

	for index := range words {
		wordLen = len(words[index])
		word := words[index] + "#" + words[index]
		newWordLen = 2*wordLen + 1

		for i := 0; i < wordLen; i++ {
			curr := &root

			for j := i; j < newWordLen; j++ {
				c = word[j]
				if curr.children[c] == nil {
					curr.children[c] = &WordFilter{children: make(map[byte]*WordFilter)}
				}

				curr = curr.children[c]
				curr.index = index
			}
		}
	}

	return root
}

func (w *WordFilter) F(pref string, suff string) int {
	word := suff + "#" + pref
	curr := w

	var c byte
	for idx := range word {
		c = word[idx]
		if curr.children[c] == nil {
			return -1
		}

		curr = curr.children[c]
	}

	return curr.index
}

/**
 * Your WordFilter object will be instantiated and called as such:
 * obj := Constructor(words);
 * param_1 := obj.F(pref,suff);
 */
